amlb/utils/process.py

```diff
@@ -83,6 +84,47 @@ def as_cmd_args(*args, **kwargs):
                        ))
 
 
+def live_output_windows(process: subprocess.Popen, **ignored) -> Tuple[str, str]:
+    """ Custom output forwarder, because select.select is not Windows compatible. """
+    def wrap_if_needed(s):
+        """ Wrap the stream in TextIO if stream produces bytes. """
+        if hasattr(process, "text_mode") and process.text_mode:
+            return s
+        return io.TextIOWrapper(s, encoding="utf-8")
+
+    outputs = dict(
+        out=dict(
+            stream=wrap_if_needed(process.stdout),
+            queue=queue.Queue(),
+            lines=[],
+        ),
+        err=dict(
+            stream=wrap_if_needed(process.stderr),
+            queue=queue.Queue(),
+            lines=[],
+        ),
+    )
+
+    def forward_output(stream, queue_):
+        for line in stream:
+            queue_.put(line)
+
+    for output in outputs.values():
+        t = threading.Thread(target=forward_output, args=(output["stream"], output["queue"]))
+        t.daemon = True
+        t.start()
+
+    while process.poll() is None:
+        for output in outputs.values():
+            try:
+                line = output["queue"].get(timeout=0.5)
```

Comment:
magic number?
I'd rather use `get_no_wait` here (this will almost always block for nothing on the `err` stream) with some kind of `ready` flag.

something like:
```python
        out=dict(
            stream=wrap_if_needed(process.stdout),
            queue=queue.Queue(),
            ready=threading.Event(),
            lines=[],
        ),

    def forward_output(stream, queue_, ready):
        for line in stream:
            queue_.put(line)
            ready.set()


while process.poll() is None and any([o['ready'].wait() for o in outputs.values()]):
   for output in outputs.values():
       with outpout['ready']._cond:
           if output['ready'].is_set():
               try: 
                    line = output['queue'].get_no_wait()
                    ...
                except queue.Empty:
                       output['ready'].clear()
```

Reply:
Wouldn't the same problem be caused by the `o['ready'].wait()` call here? By default it's blocking, but even with a `timeout`, we're waiting for the `Event` to be set on a stream which may not produce any output. I can't think of a good way around this.

While I can't think of a way to always respond to the quickest event (without rather busy waiting), you could dynamically adjust wait times: longer for streams that you expect to produce output first, shorter for those that typically don't. But I feel this is some unnecessary over-engineering. I did adjust the code to always empty a queue before starting to wait on the next one.

Regarding the magic number, two questions:
 - why is there both `timeout` and `activity_timeout`? 
 - I don't think either describe the timeout that the polling time should be (because you can rely on `select`), do you propose I make a new parameter for that?

Reply:
`timeout` is the process timeout, so you can't use this.
`activity_timeout` is more tricky and aims at detecting process inactivity: it is currently used when building docker images or running setup commands. Building an image for example can take a very long time (e.g. autoxgboost) and we want to detect if something's wrong as soon as possibly. So the `activity_timeout` (10min by default) will terminate the process if there's no output during a certain period. So you can't use this either.

> do you propose I make a new parameter for that?

No, much better to have a magic number in this case, it's just almost always a code smell indicating that something could have been done differently.

> Wouldn't the same problem be caused by the o['ready'].wait() call here? By default it's blocking, but even with a timeout, we're waiting for the Event to be set on a stream which may not produce any output. I can't think of a good way around this.

You're right. `Event` is not suitable for this. `Condition` would probably work though as `forward_output` could call `notify_all` instead of `ready.set()`.
But you can go with the timeout approach if you want something running quickly. I just like multithreading issues too much for not trying to solve them without using timeouts.

Reply:
Thanks for the help!